ophyd/ophydobj.py

```diff
@@ -34,22 +35,45 @@ class StatusBase:
     """
     This is a base class that provides a single-slot
     call back for finished.
+
+    Parameters
+    ----------
+    timeout : float, optional
+        The default timeout to use for a blocking wait, and the amount of time
+        to wait to mark the operation as failed
     """
-    def __init__(self):
+    def __init__(self, *, timeout=None):
         super().__init__()
         self._lock = RLock()
         self._cb = None
         self.done = False
         self.success = False
+        self.timeout = None
+
+        if timeout is not None:
+            self.timeout = float(timeout)
+
+        if self.timeout is not None:
+            thread = threading.Thread(target=self._timeout_thread, daemon=True)
+            self._timeout_thread = thread
+            self._timeout_thread.start()
+
+    def _timeout_thread(self):
+        '''Handle timeout'''
+        try:
+            self.wait(timeout=self.timeout,
+                      poll_rate=max(1.0, self.timeout / 10.0))
+        except TimeoutError:
+            self._finished(success=False)
+        finally:
+            self._timeout_thread = None
 
     @_locked
-    def _finished(self, *args, **kwargs):
-        # args/kwargs are not really used, but are passed.
-        # uncomment these if you want to go hunting
-        # if args:
-        #     print("this should be empty: {}".format(args))
-        # if kwargs:
-        #     print("this should be empty: {}".format(kwargs))
+    def _finished(self, success=True, **kwargs):
+        # args/kwargs are not really used, but are passed - because pyepics
```

Comment:
I thought this callback was only handed to ophyd-level callbacks, not pyepics callbacks.